proxy/router/src/lib.rs

```diff
@@ -264,3 +312,124 @@ impl<T> AsRef<T> for Reuse<T> {
+
+    enum Request {
+        NotRecognized,
+        Reusable(usize),
+        SingleUse(usize),
+    }
+
+    impl super::Recognize for Recognize {
+        type Request = Request;
+        type Response = usize;
+        type Error = ();
+        type Key = usize;
+        type RouteError = ();
+        type Service = MultiplyAndAssign;
+
+        fn recognize(&self, req: &Self::Request) -> Option<Reuse<Self::Key>> {
+            match *req {
+                Request::NotRecognized => None,
+                Request::Reusable(n) => Some(Reuse::Reusable(n)),
+                Request::SingleUse(n) => Some(Reuse::SingleUse(n)),
+            }
+        }
+
+        fn bind_service(&mut self, _: &Self::Key) -> Result<Self::Service, Self::RouteError> {
+            Ok(MultiplyAndAssign(1))
+        }
+    }
+
+    impl Service for MultiplyAndAssign {
+        type Request = Request;
+        type Response = usize;
+        type Error = ();
+        type Future = future::FutureResult<usize, ()>;
+
+        fn poll_ready(&mut self) -> Poll<(), ()> {
+            unimplemented!()
+        }
+
+        fn call(&mut self, req: Self::Request) -> Self::Future {
+            let n = match req {
+                Request::NotRecognized => unreachable!(),
+                Request::Reusable(n) => n,
+                Request::SingleUse(n) => n,
+            };
+            self.0 *= n;
+            future::ok(self.0)
+        }
+    }
+
+    impl Router<Recognize> {
+        fn call_ok(&mut self, req: Request) -> usize {
+            self.call(req).wait().expect("should route")
+        }
+
+        fn call_err(&mut self, req: Request) -> super::Error<(), ()> {
+            self.call(req).wait().expect_err("should not route")
+        }
+    }
+
+    #[test]
+    fn invalid() {
+        let mut router = Router::new(Recognize, 1);
+
+        let rsp = router.call_err(Request::NotRecognized);
+        assert_eq!(rsp, Error::NotRecognized);
+    }
+
+    #[test]
+    fn reuse_limited_by_capacity() {
+        let mut router = Router::new(Recognize, 1);
+
+        let rsp = router.call_ok(Request::Reusable(2));
+        assert_eq!(rsp, 2);
+
+        let rsp = router.call_err(Request::Reusable(3));
+        assert_eq!(rsp, Error::OutOfCapacity);
+    }
+
+    #[test]
+    fn reuse_shares_service() {
+        let mut router = Router::new(Recognize, 1);
+
+        let rsp = router.call_ok(Request::Reusable(2));
+        assert_eq!(rsp, 2);
+
+        let rsp = router.call_ok(Request::Reusable(2));
+        assert_eq!(rsp, 4);
+    }
+
+    #[test]
+    fn single_use_does_not_share_service() {
+        let mut router = Router::new(Recognize, 1);
+
+        let rsp = router.call_ok(Request::SingleUse(2));
+        assert_eq!(rsp, 2);
+
+        let rsp = router.call_ok(Request::SingleUse(2));
+        assert_eq!(rsp, 2);
+    }
+
+    #[test]
+    fn single_use_not_cached_or_limited_by_capacity() {
+        let mut router = Router::new(Recognize, 1);
```

Comment:
Take it or leave it, but I realize all these tests involve routers with capacity 1. Might be worthwhile including some tests with capacities greater than 1, as well.

Reply:
If there are some specific cases i'm missing, i'll add them. otherwise, this sounds more like a job for quickcheck (which is fine, but not my goal atm)